dev/tools/bin/find_commit.dart

```diff
@@ -0,0 +1,93 @@
+}
+
+class Commit {
+  Commit(this.hash, this.timestamp);
+
+  final String hash;
+  final DateTime timestamp;
+
+  static String formatArgument = '--format=%H %cI';
+
+  static Commit parse(String line) {
+    final int space = line.indexOf(' ');
+    return Commit(line.substring(0, space), DateTime.parse(line.substring(space+1, line.length).trimRight()));
+  }
+
+  static List<Commit> parseList(String lines) {
+    return lines.split('\n').where((String line) => line.isNotEmpty).map(parse).toList().reversed.toList();
+  }
+}
+
+String findCommit({
+  required String primaryRepoDirectory,
+  required String primaryBranch,
+  required String primaryTrunk,
+  required String secondaryRepoDirectory,
+  required String secondaryBranch,
+}) {
+  final Commit anchor;
+  if (primaryBranch == primaryTrunk) {
+    log('on $primaryTrunk, using last commit as anchor');
+    anchor = Commit.parse(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, '--max-count=1', primaryBranch]));
+  } else {
+    final List<Commit> branchCommits = Commit.parseList(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, primaryBranch]));
+    final List<Commit> trunkCommits = Commit.parseList(git(primaryRepoDirectory, <String>['log', Commit.formatArgument, primaryTrunk]));
+    if (branchCommits.isEmpty || trunkCommits.isEmpty || branchCommits.first.hash != trunkCommits.first.hash)
+      throw StateError('Branch $primaryBranch does not seem to have a common history with trunk $primaryTrunk.');
+    if (branchCommits.last.hash == trunkCommits.last.hash) {
+      log('$primaryBranch is even with $primaryTrunk, using last commit as anchor');
+      anchor = trunkCommits.last;
+    } else {
+      int index = 0;
+      while (branchCommits.length > index && trunkCommits.length > index && branchCommits[index].hash == trunkCommits[index].hash)
+        index += 1;
+      log('$primaryBranch branched from $primaryTrunk ${branchCommits.length - index} commits ago, trunk has advanced by ${trunkCommits.length - index} commits since then.');
+      anchor = trunkCommits[index - 1];
+    }
+  }
+  return git(secondaryRepoDirectory, <String>[
+    'log',
+    '--format=%H',
+    '--until=${anchor.timestamp.toIso8601String()}',
+    '--max-count=1',
+    secondaryBranch,
+  ]);
+}
+
+String git(String workingDirectory, List<String> arguments) {
+  final ProcessResult result = Process.runSync('git', arguments, workingDirectory: workingDirectory);
+  if (result.exitCode != 0 || '${result.stderr}'.isNotEmpty)
+    throw ProcessException('git', arguments, '${result.stdout}${result.stderr}', result.exitCode);
+  return '${result.stdout}';
+}
+
+void main(List<String> arguments) {
+  if (arguments.length != 1) {
+    print('This script should be run with a single argument, the path to the secondary repository.');
+  } else {
+    print(findCommit(
+      primaryRepoDirectory: '.',
+      primaryBranch: git('.', <String>['rev-parse', '--abbrev-ref', 'HEAD']).trim(),
```

Comment:
Actually, `primaryBranch` may need to be provided via a CLI argument as well, rather than calculated based on the repo. Our LUCI builds detach the HEAD of checkouts.

Reply:
As far as I can tell it works fine for detached HEADs as implemented.

Reply:
I didn't try this script, but when I run the command in my terminal:

```
~/git/flutter$ git checkout make-pub-activate-verbose
Switched to branch 'make-pub-activate-verbose'
Your branch is up to date with 'origin/make-pub-activate-verbose'.
~/git/flutter$ git rev-parse --abbrev-ref HEAD
make-pub-activate-verbose
~/git/flutter$ git checkout $(git rev-parse HEAD)
Note: switching to 'b716930a05d8784577cfee546d022af4666f29b5'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at b716930a05 add --verbose flag to pub global activate
~/git/flutter$ git rev-parse --abbrev-ref HEAD
HEAD
```

Reply:
ahh, it looks like even though `primaryBranch` resolves as `'HEAD'`, that still works as a valid git ref.

Reply:
right